api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java

```diff
@@ -210,22 +211,51 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
 				//The logic in saveObs evicts the old obs instance, so we need to update the collection
 				//with the newly loaded and voided instance, apparently reloading the encounter
 				//didn't do the tick
-				toRemove.add(o);
-				toAdd.add(os.getObs(o.getId()));
-				toAdd.add(newObs);
+				obsToRemove.add(o);
+				obsToAdd.add(os.getObs(o.getId()));
+				obsToAdd.add(newObs);
 			}
 		}
-		
-		for (Obs o : toRemove) {
+
+		removeGivenObsAndTheirGroupMembersFromEncounter(obsToRemove, encounter);
+		addGivenObsAndTheirGroupMembersToEncounter(obsToAdd, encounter);
+		return encounter;
+	}
+
+	/**
+	 * This method will remove given list of obs and their group members from encounter
+	 *
+	 * @param obsToRemove list of obs that need to be removed recursively
+	 * @param encounter the encounter from which the obs will be removed
+	 */
+	private void removeGivenObsAndTheirGroupMembersFromEncounter(List<Obs> obsToRemove, Encounter encounter) {
+		for (Obs o : obsToRemove) {
 			encounter.removeObs(o);
+			Set<Obs> groupMembers = o.getGroupMembers(true);
+			if (CollectionUtils.isNotEmpty(groupMembers)) {
+				ArrayList<Obs> membersToRemove = new ArrayList<>(groupMembers);
```

Comment:
Could you remind me as to why we instantiate a new collection on the line above?